### PR TITLE
Add loading spinner

### PR DESCRIPTION
**Pull request by jgzuke**
Designs at https://app.zeplin.io/project/5c7dcb5ab4e654bca8cde54d/screen/5d56d40acf2df541b112fc57

Spinner code mostly from https://codepen.io/daniman-the-bold/pen/jgjqBw, thanks @daniman 🎉

Also made the `DemoGroup` shared since it was being used in `TextField`, `Button`, and now `Loader` stories.

### CODE REVIEW

File: src/Loaders/Loaders.story.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                  css={{
+                    ...typography.base.small,
+                    fontWeight: 600,
+                    display: "block",
+                    textTransform: "uppercase",
+                  }}
+                >
+                  {size}
+                </span>
+                <LoadingSpinner css={{ display: "block" }} size={size} />
```
Comment (justinanastos) [inline src/Loaders/Loaders.story.tsx]: Does this `css` actually do anything considering that you don't accept a `className` prop in `LoadingSpinner`? 

I started moving towards using the more verbose and less magical `<ClassNames>{({css, cx}) => (<div className={cx(css({ display: 'block' }))} />))}</ClassNames>` because then `className`s are passed through as expected. `css` was causing me a lot of issues when trying to use `cloneElement()` (or `jsx()`, which is _not_ the same thing as `cloneElement()` because it skips `key` and `ref`s)

File: src/Loaders/Loaders.story.tsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+              </div>
+            ))}
+          </div>
+          <div
+            css={{
+              backgroundColor: colors.blilet.darker,
```
Comment (justinanastos) [inline src/Loaders/Loaders.story.tsx]: Q: Why did you chose to use `colors.blilet.darker ` instead of `colors.black.base`?

File: src/Loaders/LoadingSpinner.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  const orbitColor = darkBackground ? colors.white : colors.silver.light;
+  const orbitOpacity = darkBackground ? 0.5 : 1;
+  const asteroidColor = darkBackground ? colors.white : colors.blue.base;
+
+  const pixelSize = SIZE_MAP[size];
+
+  return (
+    <svg
+      viewBox="0 0 100 100"
+      style={{
```
Comment (justinanastos) [inline src/Loaders/LoadingSpinner.tsx]: Is there a reason you used `style` instead of emotion?

**Comment by justinanastos** [comment]
This also needs some input from @adamatorres about how to handle the disabled button state where we can't see the orbit

File: src/Loaders/LoadingSpinner.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// 3rd rotation: 240 deg and 270 deg
+// 4th rotation: 330 deg and 360 deg (restart loop)
+const SPIN = keyframes`
+  25% { transform: rotate(450deg) }
+  50% { transform: rotate(900deg) }
+  75% { transform: rotate(1350deg) }
+  100% { transform: rotate(1800deg) }
+`;
+
+const SIZE_MAP = {
```
[the remark this answers is not in the document]
Comment (jgzuke) [reply src/Loaders/LoadingSpinner.tsx]: Typescript already catches the error you mentioned above since its type stays as the specific map and doesnt get reduced to `{ [key: string]: number }`

File: src/Loaders/LoadingSpinner.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
+/** @jsx jsx */
+import React from "react";
+import { jsx, keyframes } from "@emotion/core";
+import { colors } from "../colors";
+
+export type Size = "large" | "medium" | "small" | "tiny" | "icon";
+interface Props {
```
[the remark this answers is not in the document]
Comment (jgzuke) [reply src/Loaders/LoadingSpinner.tsx]: 👍good catch, thanks

**Comment by daniman** [comment]
This is pretty great Jason, thanks for putting my pen into code here! Two quick comments on the styles:

#1 –– I think the orbit line and orbit circle probably need a bit of tweaking at the different sizes. I'm not sure what the dimension are myself.

#2 –– I think Tim had a different timing function for the orbit when it's in a button. Looks like it's a slightly faster linear animation, not a cubic bezier. You should sync up w him and Adam on the intent there: https://meteor.slack.com/archives/CFSCN3G7L/p1565744964008200.

File: src/Loaders/LoadingSpinner.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
+/** @jsx jsx */
+import React from "react";
+import { jsx, keyframes } from "@emotion/core";
+import { colors } from "../colors";
+
+export type Size = "large" | "medium" | "small" | "tiny" | "icon";
```
[the remark this answers is not in the document]
Comment (jgzuke) [reply src/Loaders/LoadingSpinner.tsx]: Sure! Im not quite clear what specifically you're suggesting though, `"large" | "medium" | "small" | "extraSmall" | "2extraSmall"`? To me it feels weird to say '2' with the full word, but shortening to `2xs` is weird when using full names for all other component sizes. One thing to note is `icon` might make it clearer which size matches the icon size for buttons etc, instead of having to try out multiple sizes

File: src/Loaders/Loaders.story.tsx (hunk 2 above)
Comment (jgzuke) [reply src/Loaders/Loaders.story.tsx]: Since I dont know of any screens with black backgrounds, black background didnt seem like a useful thing to have visualized. I have seen components with the `blilet.darker` background color like the sidebar so I thought that might be a more common background and more useful to have visualized

File: src/Loaders/Loaders.story.tsx (hunk 1 above)
Comment (jgzuke) [reply src/Loaders/Loaders.story.tsx]: `LoadingSpinner` still accepted a `className` (passing through `otherProps`). I can rewrite this using `ClassNames` if thats the way we want to go but personally I find it less readable/ergonomic compared to `css`.

File: src/Loaders/LoadingSpinner.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (jgzuke) [reply src/Loaders/LoadingSpinner.tsx]: @justinanastos I just got to @daniman s comment below about the timing changing when used as an icon. What would you think about removing `icon` as a size and adding it as its own flag? Im a bit conflicted on having multiple props affect size but if we do want to have the different animation at that size it might make sense (We can also make the type so only one of `icon` or `size` could be passed).

[the remark this answers is not in the document]
Comment (evans) [reply src/Loaders/LoadingSpinner.tsx]: I think `2xs` stands for `extraExtraSmall` and is more readable than `xxs`, especially as things get more `extra` 😆. I'm into keeping the names expanded, since users can add their own semantics on top of these if they want. `2extraSmall` makes sense to me. I think the animation changes depending on the size are a space-kit concern. Their use is an application concern, so I think `icon` should not show up in the api

[the remark this answers is not in the document]
Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: This suggestion is coming from this comment by @timbotnik :

https://github.com/mdg-private/engine-frontend/pull/1606#pullrequestreview-277500719

> This will be great, however I would prefer if we kept parallel with the Tailwind text naming conventions? For example, instead of `sk-text-small` we should use `sk-text-sm`, instead of `sk-text-xxlarge` we should use `sk-text-2xl`, etc.
> 
> > .text-xs | font-size: .75rem;
> > .text-sm | font-size: .875rem;
> > .text-base | font-size: 1rem;
> > .text-lg | font-size: 1.125rem;
> > .text-xl | font-size: 1.25rem;
> > .text-2xl | font-size: 1.5rem;
> > .text-3xl | font-size: 1.875rem;
> > .text-4xl | font-size: 2.25rem;
> > .text-5xl | font-size: 3rem;
> > .text-6xl | font-size: 4rem;

We've been using `xs`, `sm`, `base`, `lg`, `xl`, `2xl`, `3xl`, etc for a year and the language feels right to me. This is subjective, and we're talking about bringing something I personally feel very comfortable with in graph manager to Space Kit. I want to point out that this is *not* for the sake of consistency; this is for the sake of I think this is a good API.

File: src/Loaders/Loaders.story.tsx (hunk 2 above)
Comment (justinanastos) [reply src/Loaders/Loaders.story.tsx]: Great answer. We're going to have pages with black backgrounds in the near future. 

Check out this quip doc: https://apollographql.quip.com/ZQtPAw13fRcH/Information-Architecture-Nav-Concepts#RMBACASwNB7

Also the background on buttons: https://zpl.io/amdN6Pr

TLDR; Please change the background color to `colors.black.base`

File: src/Loaders/LoadingSpinner.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: Well that's not exactly true. TypeScript does not catch the above error when we add a new `Size` and forget to add it to `SIZE_MAP`. Somewhere else in the code will have an error, but that's only because you chose to not enforce your typing where it matters. In other words, there is a TypeScript error, but that's by a happy coincidence. I'm asking you to make `SIZE_MAP` explicitly defined as "this is a number for every `Size`", which is _not_ what you're doing now. This will show exactly what our intent was.

## Before

```diff
diff --git a/src/Loaders/LoadingSpinner.tsx b/src/Loaders/LoadingSpinner.tsx
index a9632aa..6279c5b 100644
--- a/src/Loaders/LoadingSpinner.tsx
+++ b/src/Loaders/LoadingSpinner.tsx
@@ -3,7 +3,7 @@ import React from "react";
 import { jsx, keyframes } from "@emotion/core";
 import { colors } from "../colors";
 
-export type Size = "large" | "medium" | "small" | "tiny" | "icon";
+export type Size = "huge" | "large" | "medium" | "small" | "tiny" | "icon";
 export type Theme = "light" | "dark";
 interface Props {
   /**
```

```shell
➜  space-kit git:(jgzuke/add-loading-spinner) ✗ npx tsc --noEmit
src/Loaders/LoadingSpinner.tsx:68:21 - error TS7053: Element implicitly has an 'any' type because expression of type 'Size' can't be used to index type '{ large: number; medium: number; small: number; tiny: number; icon: number; }'.
  Property 'huge' does not exist on type '{ large: number; medium: number; small: number; tiny: number; icon: number; }'.

68   const pixelSize = SIZE_MAP[size];
                       ~~~~~~~~~~~~~~


Found 1 error.
```

## After

This is in contrast to making the requested change, which will show exactly what we're expecting instead of relying on an implicit error elsewhere:

```diff
diff --git a/src/Loaders/LoadingSpinner.tsx b/src/Loaders/LoadingSpinner.tsx
index a9632aa..e306927 100644
--- a/src/Loaders/LoadingSpinner.tsx
+++ b/src/Loaders/LoadingSpinner.tsx
@@ -3,7 +3,7 @@ import React from "react";
 import { jsx, keyframes } from "@emotion/core";
 import { colors } from "../colors";
 
-export type Size = "large" | "medium" | "small" | "tiny" | "icon";
+export type Size = "huge" | "large" | "medium" | "small" | "tiny" | "icon";
 export type Theme = "light" | "dark";
 interface Props {
   /**
@@ -36,7 +36,7 @@ const SPIN = keyframes`
   100% { transform: rotate(1800deg) }
 `;
 
-const SIZE_MAP = {
+const SIZE_MAP: Record<Size, number> = {
   large: 90,
   medium: 64,
   small: 48,

```

```zsh
➜  space-kit git:(jgzuke/add-loading-spinner) ✗ npx tsc --noEmit
src/Loaders/LoadingSpinner.tsx:39:7 - error TS2741: Property 'huge' is missing in type '{ large: number; medium: number; small: number; tiny: number; icon: number; }' but required in type 'Record<Size, number>'.

39 const SIZE_MAP: Record<Size, number> = {
         ~~~~~~~~


Found 1 error.
```

File: src/Loaders/LoadingSpinner.tsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+const SIZE_MAP = {
+  large: 90,
+  medium: 64,
+  small: 48,
+  tiny: 32,
+  icon: 16,
+};
+
+const THEME_MAP = {
```
Comment (justinanastos) [inline src/Loaders/LoadingSpinner.tsx]: I'd like this to be strictly typed based on the discussion in the same PR.

```diff
diff --git a/src/Loaders/LoadingSpinner.tsx b/src/Loaders/LoadingSpinner.tsx
index a9632aa..fd3be28 100644
--- a/src/Loaders/LoadingSpinner.tsx
+++ b/src/Loaders/LoadingSpinner.tsx
@@ -1,4 +1,5 @@
 /** @jsx jsx */
+import * as CSS from "csstype";
 import React from "react";
 import { jsx, keyframes } from "@emotion/core";
 import { colors } from "../colors";
@@ -44,7 +45,14 @@ const SIZE_MAP = {
   icon: 16,
 };
 
-const THEME_MAP = {
+const THEME_MAP: Record<
+  Theme,
+  {
+    orbitColor: CSS.ColorProperty;
+    orbitOpacity: CSS.GlobalsNumber;
+    asteroidColor: CSS.ColorProperty;
+  }
+> = {
   light: {
     orbitColor: colors.silver.light,
     orbitOpacity: 1,

```

File: src/Loaders/Loaders.story.tsx (hunk 1 above)
Comment (justinanastos) [reply src/Loaders/Loaders.story.tsx]: I wasn't suggesting `ClassNames`, just mentioning that I had to use it elsewhere. I also find it less ergonomic, but sometimes it's the only way to get things done.

I missed that you were passing through the other props. 👍

File: src/Loaders/LoadingSpinner.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (evans) [reply src/Loaders/LoadingSpinner.tsx]: 🤔 I'm comfortable with either as long as the names are consistent within space kit, ie

https://github.com/apollographql/space-kit/blob/16a47a1017b4d5c512e3e56406331993ef0309cb/src/Button/Button.tsx#L172

I lean towards full english for space-kit, since the sizes stand on their own. I've found short names with tailwinds are useful and necessary, since we write so many of them. Because they are constant throughout our application the language/effects are top of mind with little recall effort and brevity means efficiency. I believe that these components will be used once or twice per component/file, so brevity is less important and expressiveness is more. That's my opinion. I'll be happy either way and defer to your decision. Look forward to learning/seeing it used 🎉

[the remark this answers is not in the document]
Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: I can handle that @evans . I do not want to have ambiguity though:

```jsx
// I don't know if `tiny` or `icon` is smaller. 
"large" | "medium" | "small" | "tiny" | "icon";
```

How about this:
```jsx
"large" | "medium" | "small" | "xsmall" | "2xsmall";
```

File: src/Loaders/LoadingSpinner.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: I didn't say this in the comment so there's no fault on your part; I was hoping you'd respond to this message instead of just making the change and requesting a new review. 

It's important to me that you understand why I'm requesting this change and why I think this makes future code changes safer. Why? If you are looking for the same things, then you'll catch places where I might have missed strong type safety in my PRs. A rising tide lifts all ships!

File: src/Loaders/LoadingSpinner.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: I can handle that @evans . I do not want to have ambiguity though:

```jsx
// I don't know if `tiny` or `icon` is smaller. 
"large" | "medium" | "small" | "tiny" | "icon";
```

How about this:
```jsx
"large" | "medium" | "small" | "xsmall" | "2xsmall";
```

File: src/Loaders/Loaders.story.tsx (hunk 2 above)
Comment (justinanastos) [reply src/Loaders/Loaders.story.tsx]: Great answer. We're going to have pages with black backgrounds in the near future. 

Check out this quip doc: https://apollographql.quip.com/ZQtPAw13fRcH/Information-Architecture-Nav-Concepts#RMBACASwNB7

Also the background on buttons: https://zpl.io/amdN6Pr

TLDR; Please change the background color to `colors.black.base`

File: src/Loaders/LoadingSpinner.tsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+  /**
+   * Theme for the spinner
+   */
+  theme?: Theme;
+
+  /**
+   * Size of the spinner
+   */
+  size?: Size;
```
Comment (justinanastos) [inline src/Loaders/LoadingSpinner.tsx]: First, this isn't in the README, but it should be!

Can you add the default value here so when we have automatic prop table generation (hopefully in a few working days, this is a new priority along with other docs work), then the default will be appropriately listed?

Example: https://github.com/strothj/react-docgen-typescript-loader#exporting-components

Same for all other defaulted props please (meaning `theme`).

File: src/Loaders/LoadingSpinner.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: I didn't say this in the comment so there's no fault on your part; I was hoping you'd respond to this message instead of just making the change and requesting a new review. 

It's important to me that you understand why I'm requesting this change and why I think this makes future code changes safer. Why? If you are looking for the same things, then you'll catch places where I might have missed strong type safety in my PRs. A rising tide lifts all ships!

File: src/Loaders/LoadingSpinner.tsx (hunk 8 above)
Comment (jgzuke) [reply src/Loaders/LoadingSpinner.tsx]: Thats really cool, I didnt know about that!

Comment (justinanastos) [reply src/Loaders/LoadingSpinner.tsx]: I didn't either :) None of the `Button`s props have the `@default` yet 🙈

**Comment by justinanastos** [comment]
Get eslint to pass (the error is auto-fixable, looks like it's time to add a global vs code extension that auto fixes things).

**Comment by justinanastos** [comment]
:rocket: PR was released in v1.8.0 :rocket:
